### PR TITLE
perf: cache file stats/reads

### PR DESCRIPTION
**Pull request by keithamus**
This is some work towards fixing (not entirely) #121. File reads and stat calls are cached for the lifetime of the code execution. This takes the number of times files are stated dramatically down (on our build, one file goes from 13,000 stat calls to 1,300), however `browser-resolve` still does many file reads/stats using its own mechanism - so further PRs will be arriving next week around this, as files are still read multiple times when they don't need to be. 

It may be worth clearing the read file cache out on an `ongenerate` rollup call or something? 

Thoughts/comments welcome.

/cc @mislav who worked on this with me

### CODE REVIEW

**Comment by keithamus** [comment]
Yes the other issue is if a file gets into an error state, then that error state is cached indefinitely. I think I'll address some of these concerns though as Im not entirely comfortable merging this PR in this state.

**Comment by keithamus** [comment]
@lukastaegert would be good to get your thoughts about this new change. I've changed two things since last review:

 - If the stat/read calls error (except if a stat call occurs on a non-existent file) then delete the cache, so they are re-read/stated each time.
 - The plugin adds the `write` hook (per https://github.com/rollup/rollup/wiki/Plugins#creating-plugins) to flush the caches every time a compile is run.

**Comment by lukastaegert** [comment]
Thanks for your efforts here! Would be nice if you could rebase against current master as this will add a "prepare" script that should enable direct installation from Github. This should make it easier to test this PR in different scenarios.

**Comment by keithamus** [comment]
Rebased 😸

File: src/index.js [hunk 1]
```diff
@@ -8,6 +8,37 @@ const ES6_BROWSER_EMPTY = resolve( __dirname, '../src/empty.js' );
...
+let isFileCache = {};
+function cachedIsFile (file, cb) {
+	if (file in isFileCache === false) {
+		isFileCache[file] = statAsync(file)
+			.then(
+				stat => stat.isFile(),
+				err => {
+					if (err.code == 'ENOENT') return false;
+					throw err;
+				}
```
Comment (lukastaegert) [inline src/index.js]: Why not just put this into the catch block below? Also, using *both* the second `then` argument *and* `catch` seems a little over the top to me... 😉

File: src/index.js [hunk 2]
```diff
@@ -31,6 +62,11 @@ export default function nodeResolve ( options = {} ) {
 	return {
 		name: 'node-resolve',
 
+		write () {
+			isFileCache = {};
+			readFileCache = {};
+		},
+
```
Comment (lukastaegert) [inline src/index.js]: Very good idea but you probably mean `onwrite`

**Comment by keithamus** [comment]
Fixed up those issues. I'm comfortable enough with this now. We've been testing this internally for a while now and it works well.

**Comment by lukastaegert** [comment]
Hey @keithamus, sorry for not getting back to you. So it seems that while I do not have the npm publishing rights for this plugin, *you*  do have them. So I was wondering if you would want to take it upon you to release the latest PRs for this plugin?
I'm also asking because @kellyselden was asking me about the state of #140 and #141.

**Comment by keithamus** [comment]
For sure - are you happy with each of these PRs?

**Comment by lukastaegert** [comment]
Yup, thanks for taking care of this 👍! This really helps @guybedford and myself concentrate on maintaining rollup core. If you want a review from us for anything, don't hesitate to add us as reviewers.

**Comment by lukastaegert** [comment]
By the way, at the moment I am fledging out a solution to get precise performance timings for rollup both for internal code as well as time spend in plugin hooks rollup/rollup#2045.

When bundling rollup with rollup-plugin-node-resolve@3.0.0, rollup spends about 180ms resolving ids on my machine. With rollup-plugin-node-resolve@3.2.0, this drops to 4ms! Amazing work indeed! 🎉

**Pull request merged**